File: adapters/dotnet/src/AuthNinja.AspNetCore/Auth/FieldEncryption.cs
using System.Security.Cryptography;
using System.Text;

namespace AuthNinja.AspNetCore.Auth;

/// <summary>AES-256-GCM field encryption keyed from AUTH_NINJA_SECRET.</summary>
internal static class FieldEncryption
{
    private const string Version = "v1";
    private const int IvLength = 12;
    private const int KeyLength = 32;
    private const string ScryptSalt = "auth-ninja-field-encryption";

    public static string Encrypt(string plaintext, string secret)
    {
        if (plaintext.Length == 0)
        {
            throw new InvalidOperationException("Cannot encrypt empty field value");
        }

        var key = DeriveKey(secret);
        var iv = RandomNumberGenerator.GetBytes(IvLength);
        var plaintextBytes = Encoding.UTF8.GetBytes(plaintext);
        var ciphertext = new byte[plaintextBytes.Length];
        var tag = new byte[16];

        using var aes = new AesGcm(key, tag.Length);
        aes.Encrypt(iv, plaintextBytes, ciphertext, tag);

        var payload = new byte[iv.Length + ciphertext.Length + tag.Length];
        Buffer.BlockCopy(iv, 0, payload, 0, iv.Length);
        Buffer.BlockCopy(ciphertext, 0, payload, iv.Length, ciphertext.Length);
        Buffer.BlockCopy(tag, 0, payload, iv.Length + ciphertext.Length, tag.Length);

        return $"{Version}:{Base64UrlHelper.Encode(payload)}";
    }

    public static string Decrypt(string stored, string secret)
    {
        if (!stored.StartsWith($"{Version}:", StringComparison.Ordinal))
        {
            return stored;
        }

        var payload = Base64UrlHelper.Decode(stored[(Version.Length + 1)..]);
        if (payload.Length <= IvLength + 16)
        {
            throw new InvalidOperationException("Invalid encrypted field payload");
        }

        var iv = payload.AsSpan(0, IvLength);
        var tag = payload.AsSpan(payload.Length - 16, 16);
        var ciphertext = payload.AsSpan(IvLength, payload.Length - IvLength - 16);
        var plaintext = new byte[ciphertext.Length];

        using var aes = new AesGcm(DeriveKey(secret), tag.Length);
        aes.Decrypt(iv, ciphertext, tag, plaintext);

        return Encoding.UTF8.GetString(plaintext);
    }

    private static byte[] DeriveKey(string secret) =>
        Rfc2898DeriveBytes.Pbkdf2(
            secret,
            Encoding.UTF8.GetBytes(ScryptSalt),
            100_000,
            HashAlgorithmName.SHA256,
            KeyLength);
}

File: packages/core/src/field-encryption.ts
import { createCipheriv, createDecipheriv, randomBytes, scryptSync } from "node:crypto";

const FIELD_ENCRYPTION_VERSION = "v1";
const ALGORITHM = "aes-256-gcm";
const IV_LENGTH = 12;
const KEY_LENGTH = 32;
const SCRYPT_SALT = "auth-ninja-field-encryption";

function deriveKey(secret: string): Buffer {
  return scryptSync(secret, SCRYPT_SALT, KEY_LENGTH);
}

/** Encrypt a sensitive field (e.g. TOTP seed) with AES-256-GCM keyed from AUTH_NINJA_SECRET. */
export function encryptField(plaintext: string, secret: string): string {
  if (plaintext.length === 0) {
    throw new Error("Cannot encrypt empty field value");
  }

  const key = deriveKey(secret);
  const iv = randomBytes(IV_LENGTH);
  const cipher = createCipheriv(ALGORITHM, key, iv);
  const encrypted = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  const payload = Buffer.concat([iv, encrypted, tag]);
  return `${FIELD_ENCRYPTION_VERSION}:${payload.toString("base64url")}`;
}

/**
 * Decrypt a field encrypted with {@link encryptField}.
 * Plaintext legacy values (pre-encryption) are returned unchanged for migration.
 */
export function decryptField(stored: string, secret: string): string {
  if (!stored.startsWith(`${FIELD_ENCRYPTION_VERSION}:`)) {
    return stored;
  }

  const payload = Buffer.from(stored.slice(FIELD_ENCRYPTION_VERSION.length + 1), "base64url");
  if (payload.length <= IV_LENGTH + 16) {
    throw new Error("Invalid encrypted field payload");
  }

  const iv = payload.subarray(0, IV_LENGTH);
  const tag = payload.subarray(payload.length - 16);
  const ciphertext = payload.subarray(IV_LENGTH, payload.length - 16);

  const key = deriveKey(secret);
  const decipher = createDecipheriv(ALGORITHM, key, iv);
  decipher.setAuthTag(tag);
  const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  return decrypted.toString("utf8");
}
